hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/model/HoodieFlinkRecord.java

```diff
@@ -112,15 +113,24 @@ public HoodieRecord joinWith(HoodieRecord other, Schema targetSchema) {
 
   @Override
   public HoodieRecord prependMetaFields(Schema recordSchema, Schema targetSchema, MetadataValues metadataValues, Properties props) {
-    int metaFieldSize = targetSchema.getFields().size() - recordSchema.getFields().size();
-    GenericRowData metaRow = new GenericRowData(metaFieldSize);
+    boolean withMetaFields = recordSchema.getField(RECORD_KEY_METADATA_FIELD) != null;
+    boolean withOperationField = targetSchema.getField(OPERATION_METADATA_FIELD) != null;
+    int metaFieldSize = HOODIE_META_COLUMNS.size();
+    String[] metaFields = new String[metaFieldSize];
+    if (withMetaFields) {
+      for (int i = 0; i < metaFieldSize; i++) {
+        metaFields[i] = data.getString(i).toString();
+      }
+    }
+
+    AbstractHoodieRowData rowWithMetaFields = HoodieRowDataCreation.create(metaFields, data, withOperationField, withMetaFields);
```

Comment:
Should we just store `StringData` for `AbstractHoodieRowData` metadata fields.

Reply:
We can do that for memory efficiency.